src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveNotificationListener.cs

```diff
@@ -56,190 +58,141 @@ internal class FileMoveNotificationListener : IFileMoveNotificationListener
-            {
-                filesToMove = null;
-
-                foreach (IFileMoveItem item in items)
+                Document? currentDocument = project.Documents.FirstOrDefault(d => StringComparers.Paths.Equals(d.FilePath, itemToMove.Source));
+                if (currentDocument is null)
                 {
-                    RecursiveTryGetFilesToMove(item, ref filesToMove);
+                    continue;
                 }
 
-                return filesToMove is not null;
-            }
+                // Get the relative folder path from the project to the destination.
+                string destinationFolderPath = Path.GetDirectoryName(_unconfiguredProject.MakeRelative(itemToMove.Destination));
+                string[] destinationFolders = destinationFolderPath.Split(Delimiter.Path, StringSplitOptions.RemoveEmptyEntries);
 
-            void RecursiveTryGetFilesToMove(IFileMoveItem? item, ref List<(string file, string destination)>? filesToMove)
-            {
-                if (item is null)
+                // Since this rename only moves the location of the file to another directory, it will use the SyncNamespaceDocumentAction in Roslyn as the rename action within this set.
+                // The logic for selecting this rename action can be found here: https://github.com/dotnet/roslyn/blob/960f375f4825a189937d4bfd9fea8162ecc63177/src/Workspaces/Core/Portable/Rename/Renamer.cs#L133-L136
+                RenameDocumentActionSet renameActionSet = await RenameDocumentAsync(currentDocument, s_renameOptions, null, destinationFolders);
+                if (renameActionSet.ApplicableActions.IsEmpty || renameActionSet.ApplicableActions.Any(aa => aa.GetErrors().Any()))
                 {
-                    return;
+                    continue;
                 }
 
-                if (item.IsFolder)
-                {
-                    if (item is not ICopyPasteItem copyPasteItem)
-                    {
-                        return;
-                    }
-
-                    foreach (IFileMoveItem child in copyPasteItem.Children.Cast<IFileMoveItem>())
-                    {
-                        RecursiveTryGetFilesToMove(child, ref filesToMove);
-                    }
-                }
-                else
-                {
-                    bool isCompileItem = StringComparers.ItemTypes.Equals(item.ItemType, Compile.SchemaName);
+                // Getting the rename message requires an instance of RenameDocumentAction.
+                // We only need to set this message text once for the lifetime of the class, since it isn't dynamic.
+                // Even though it isn't dynamic, it does get localized appropriately in Roslyn.
+                // The text in English is "Sync namespace to folder structure".
+                _renameMessage ??= renameActionSet.ApplicableActions.First().GetDescription();
```

Comment:
Are you sure that `First` won't throw? Can this sequence never be empty?

Reply:
I didn't write this expression initially. Looking into it, here is the code that produces the actions:
https://github.com/dotnet/roslyn/blob/50e8196bc8fa8f227bc65bd32917740e59a49812/src/Workspaces/Core/Portable/Rename/Renamer.cs#L125-L148

The first parameter of `new RenameDocumentActionSet()` is what gets set to the `ApplicableActions` property. Based on the if-statements, I don't think it is physically possible to get to this point in CPS and have those if-statements fail, since this only activates when moving a file... which would **always** cause the second if-statement about document folders to be `true`.

Even if I did `FirstOrDefault` instead, what would be appropriate default? Empty string?